src/MouseMoveProvider.js

```diff
@@ -0,0 +1,75 @@
+// @flow
+import React, { Component, PropTypes } from 'react';
+import { uuid } from './utils';
+import invariant from 'invariant';
+
+type Subscription = { uuid: string, callback: () => void };
+type SubscriptionArray = [Subscription];
+
+export default class MouseMoveProvider extends Component {
+  subs: SubscriptionArray = [];
+  hasSubs: Boolean = false;
+  initialOffset: Number = window.pageYOffset;
+
+  static propTypes = {
+    children: PropTypes.node
+  };
+
+  static childContextTypes = {
+    mouseMove: PropTypes.object.isRequired
+  };
+
+  getChildContext() {
+    const { subscribe } = this;
+    return {
+      mouseMove: {
+        subscribe
+      }
+    };
+  }
+
+  componentWillMount() {
+    const { children } = this.props;
+    invariant(children === null || React.Children.count(children) === 1, 'A <MouseMoveProvider> may have only one child element');
+  }
+
+  componentWillUnmount() {
+    this.subs = [];
+    this.removeListener();
+  }
+
+  subscribe = callback => {
+    const sub_id: string = uuid();
+    this.subs = this.subs.concat([{ sub_id, callback }]);
+    if (!this.hasSubs) {
+      this.hasSubs = true;
+      this.registerListener();
+    }
+    return () => {
+      this.subs = this.subs.filter((sub: Subscription) => sub.uuid !== sub_id);
+      if (this.subs.length === 0) {
+        this.hasSubs = false;
+        this.removeListener();
+      }
+    };
+  };
```

Comment:
I'd like to avoid having to generate an id and allocating an object for each callback. I think we can simply this if we expose an `unsubscribe` method via context that takes the callback to unsubscribe as its argument.

Then we simplify this to:

```js
subscribe = (callback) => {
  this.subscriptions.push(callback)
}

unsubscribe = (callback) => {
  this.subscriptions = this.subscriptions.filter(fn => fn !== callback);
}

```

Then `Perimeter` just calls `unsubscribe(this.handleMouseMove)` in `removeEventListeners`. It also means we don't have to create an unsubscribe function for each subscription.